p2p/peer.go

```diff
@@ -12,118 +12,166 @@ import (
 	"runtime/debug"
 	"sort"
 	"strconv"
+	"strings"
 	"sync"
 	"time"
 
+	"github.com/gogo/protobuf/proto"
+	"github.com/google/orderedcode"
+	dbm "github.com/tendermint/tm-db"
+
 	"github.com/tendermint/tendermint/libs/cmap"
 	"github.com/tendermint/tendermint/libs/log"
 	"github.com/tendermint/tendermint/libs/service"
 	tmconn "github.com/tendermint/tendermint/p2p/conn"
+	p2pproto "github.com/tendermint/tendermint/proto/tendermint/p2p"
 )
 
-// PeerAddress is a peer address URL.
+// PeerAddress is a peer address URL. If the URL is opaque, i.e. of the form
+// "scheme:<opaque>", then the opaque part has to contain either the node ID
+// or a node ID and path in the form "scheme:<nodeid>@<path>".
+//
+// PeerAddress differs from Endpoint in that the address hostname may be
+// expanded into multiple IP addresses (thus multiple endpoints).
 type PeerAddress struct {
-	*url.URL
-}
+	ID       NodeID
+	Protocol Protocol
+	Hostname string
+	Port     uint16
+	Path     string
+}
+
+// ParsePeerAddress parses a peer address URL into a PeerAddress,
+// normalizing and validating it.
+func ParsePeerAddress(urlString string) (PeerAddress, error) {
+	url, err := url.Parse(urlString)
+	if err != nil || url == nil {
+		return PeerAddress{}, fmt.Errorf("invalid peer address %q: %w", urlString, err)
+	}
+
+	address := PeerAddress{}
+
+	// If the URL is opaque, i.e. in the form "scheme:<opaque>", we specify the
+	// opaque bit to be either a node ID or a node ID and path in the form
+	// "scheme:<nodeid>@<path>".
+	if url.Opaque != "" {
+		parts := strings.Split(url.Opaque, "@")
+		if len(parts) > 2 {
+			return PeerAddress{}, fmt.Errorf("invalid address format %q, unexpected @", urlString)
+		}
+		address.ID, err = NewNodeID(parts[0])
+		if err != nil {
+			return PeerAddress{}, fmt.Errorf("invalid peer ID %q: %w", parts[0], err)
+		}
+		if len(parts) == 2 {
+			address.Path = parts[2]
```

Comment:
```suggestion
			address.Path = parts[1]
```

Reply:
Thanks! I see the LGTM-bot was trying to warn about this as well, good bot.